### PR TITLE
Fix no space left error in CI environment

### LINKED ISSUE #3408: E2E test fails with error: no space left on device
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

```
Warning  Failed     18m                  kubelet            Failed to pull image "nvcr.io/nvidia/tritonserver:23.05-py3": rpc error: code = Unknown desc = failed to register layer: write /opt/tritonserver/backends/tensorflow/libtensorflow_cc.so.2: no space left on device
```
Huggingface server python test
```
• Installing virtualenv (20.21.0)
  • Installing websocket-client (1.7.0)

  OSError

  [Errno 28] No space left on device

  at /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/poetry/utils/helpers.py:159 in download_with_progress
      [155](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:156)│         fetched_size = 0
      [156](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:157)│         with atomic_open(self._dest) as f:
      [157](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:158)│             for chunk in self._response.iter_content(chunk_size=chunk_size):
      [158](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:159)│                 if chunk:
    → [159](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:160)│                     f.write(chunk)
      [160](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:161)│                     fetched_size += len(chunk)
      [161](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:162)│                     yield fetched_size
      [162](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:163)│ 
      [163](https://github.com/kserve/kserve/actions/runs/7775049490/job/21200559350?pr=3407#step:35:164)│ 

The following error occurred when trying to handle this error:


  OSError

  [Errno 28] No space left on device

  at /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/requests/utils.py:306 in atomic_open
       302│     """Write a file to the disk in an atomic fashion"""
       303│     tmp_descriptor, tmp_name = tempfile.mkstemp(dir=os.path.dirname(filename))
       304│     try:
       305│         with os.fdopen(tmp_descriptor, "wb") as tmp_handler:
    →  306│             yield tmp_handler
       307│         os.replace(tmp_name, filename)
       308│     except BaseException:
       309│         os.remove(tmp_name)
       310│         raise

Cannot install torch.

make: *** [Makefile:5: install_dependencies] Error 1
```

Reported disk usage:
```
Filesystem     Type  1K-blocks     Used Available Use% Mounted on
  /dev/root      ext4   76026616 73270916   2739316  97% /
  tmpfs          tmpfs   8182512      172   8182340   1% /dev/shm
  tmpfs          tmpfs   3273008     6916   3266092   1% /run
  tmpfs          tmpfs      5120        0      5120   0% /run/lock
  /dev/sda[15](https://github.com/kserve/kserve/actions/runs/7775049489/job/21200559524?pr=3407#step:20:16)     vfat     106832     6186    100646   6% /boot/efi
  /dev/sdb1      ext4   76829444  4194336  68686668   6% /mnt
  tmpfs          tmpfs   [16](https://github.com/kserve/kserve/actions/runs/7775049489/job/21200559524?pr=3407#step:20:17)36500       12   1636488   1% /run/user/1001
```

**What did you expect to happen:**

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version:
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:
Available storage after cleanup at the beginning:
```
Filesystem      Size  Used Avail Use% Mounted on
/dev/root        73G   36G   38G  49% /
tmpfs           7.9G  172K  7.9G   1% /dev/shm
tmpfs           3.2G  1.1M  3.2G   1% /run
tmpfs           5.0M     0  5.0M   0% /run/lock
/dev/sdb15      105M  6.1M   99M   6% /boot/efi
/dev/sda1        74G  4.1G   66G   6% /mnt
tmpfs           1.6G   12K  1.6G   1% /run/user/1001
```
It seems root volume storage is not enough even after cleaning up the unnecessary dependencies that come with github runner.

We can utilize the /mnt volume which has plenty of storage space. One simple way is to move the docker root data directory to /mnt volume, since most of the storage usage comes from the docker images.
- This PR creates a symlink to /var/lib/docker -> /mnt/docker.
- In python tests huggingfaceserver tests also failing with no space left error. To resolve this, the dependencies of huggingfaceserver are installed in /mnt/python/huggingfaceserver-venv.
- Change poetry cache directory to /mnt/poetry for python tests
 

Disk usage after the tests completed:
- python tests:

Before the fix:
```
Filesystem     Type   Size  Used Avail Use% Mounted on
/dev/root      ext4    73G   70G  3.5G  96% /
tmpfs          tmpfs  7.9G  172K  7.9G   1% /dev/shm
tmpfs          tmpfs  3.2G  1.1M  3.2G   1% /run
tmpfs          tmpfs  5.0M     0  5.0M   0% /run/lock
/dev/sda15     vfat   105M  6.1M   99M   6% /boot/efi
/dev/sdb1      ext4    74G  9.1G   61G  14% /mnt
tmpfs          tmpfs  1.6G   [12](https://github.com/kserve/kserve/actions/runs/7795869189/job/21259533757#step:40:13)K  1.6G   1% /run/user/1001
```
After the fix:
```
Filesystem     Type   Size  Used Avail Use% Mounted on
/dev/root      ext4    73G   43G   31G  59% /
tmpfs          tmpfs  7.9G  172K  7.9G   1% /dev/shm
tmpfs          tmpfs  3.2G  1.1M  3.2G   1% /run
tmpfs          tmpfs  5.0M     0  5.0M   0% /run/lock
/dev/sdb15     vfat   105M  6.1M   99M   6% /boot/efi
/dev/sda1      ext4    74G   15G   56G  21% /mnt
tmpfs          tmpfs  1.6G   [12](https://github.com/kserve/kserve/actions/runs/7796971517/job/21262661619?pr=3411#step:41:13)K  1.6G   1% /run/user/1001
```
- test-fast e2e tests:

Before the fix:
```
Filesystem     Type  1K-blocks     Used Available Use% Mounted on
  /dev/root      ext4   76026616 73270916   2739316  97% /
  tmpfs          tmpfs   8182512      172   8182340   1% /dev/shm
  tmpfs          tmpfs   3273008     6916   3266092   1% /run
  tmpfs          tmpfs      5120        0      5120   0% /run/lock
  /dev/sda[15](https://github.com/kserve/kserve/actions/runs/7775049489/job/21200559524?pr=3407#step:20:16)     vfat     106832     6186    100646   6% /boot/efi
  /dev/sdb1      ext4   76829444  4194336  68686668   6% /mnt
  tmpfs          tmpfs   [16](https://github.com/kserve/kserve/actions/runs/7775049489/job/21200559524?pr=3407#step:20:17)36500  
```

After the fix:
```
Filesystem     Type   Size  Used Avail Use% Mounted on
  /dev/root      ext4    73G   42G   32G  58% /
  tmpfs          tmpfs  7.9G  172K  7.9G   1% /dev/shm
  tmpfs          tmpfs  3.2G  6.0M  3.2G   1% /run
  tmpfs          tmpfs  5.0M     0  5.0M   0% /run/lock
  /dev/sda[15](https://github.com/kserve/kserve/actions/runs/7796153221/job/21261046458#step:20:16)     vfat   105M  6.1M   99M   6% /boot/efi
  /dev/sdb1      ext4    74G   34G   36G  49% /mnt
  tmpfs          tmpfs  1.6G   12K  1.6G   1% /run/user/1001
```
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3408

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Any luck? I just noticed this issue as well, e.g. https://github.com/kserve/kserve/actions/runs/7790818983/job/21245544935?pr=3395

File: .github/workflows/python-test.yml [hunk 1]
```diff
@@ -225,24 +225,33 @@ jobs:
...
+          sudo chown -R $USER /mnt/python/huggingfaceserver-venv
           cd python/huggingfaceserver
           make install_dependencies
       - name: Install huggingface server
+        if: ${{ !startsWith(steps.setup-python.outputs.python-version, '3.8') }}
         run: |
           cd python/huggingfaceserver
           make dev_install
       - name: Test huggingfaceserver
+        if: ${{ !startsWith(steps.setup-python.outputs.python-version, '3.8') }}
```
Comment (sivanantha321) [inline .github/workflows/python-test.yml:266]: huggingfaceserver does not support python 3.8 ? https://github.com/kserve/kserve/blob/0b2c7d9a9fa6d766adf02449f1c12d492c62a3f1/python/huggingfaceserver/pyproject.toml#L13

**Comment by sivanantha321** [comment]
> Any luck? I just noticed this issue as well, e.g. https://github.com/kserve/kserve/actions/runs/7790818983/job/21245544935?pr=3395

Should be fine once this is merged.

**Comment by sivanantha321** [comment]
@yuzisun Can you help review the PR ?

File: .github/workflows/python-test.yml (hunk 1 above)
Comment (yuzisun) [reply .github/workflows/python-test.yml:266]: Also 3.8 is eol on oct 2024

File: .github/actions/free-up-disk-space/action.yml [hunk 2]
```diff
@@ -23,12 +23,30 @@ runs:
...
-        df -h
+        df -hT
+
+    - name: Move docker data directory
+      shell: bash
+      run: |
+        echo "Stopping docker service ..."
+        sudo systemctl stop docker
+        DOCKER_DEFAULT_ROOT_DIR=/var/lib/docker
+        DOCKER_ROOT_DIR=/mnt/docker
```
Comment (yuzisun) [inline .github/actions/free-up-disk-space/action.yml:42]: I wonder if there are better ways to configure beforehand without restarting docker

Comment (yuzisun) [reply .github/actions/free-up-disk-space/action.yml:42]: Discussed with @sivanantha321, configuring daemon.json does not work with minikube. Also https://github.com/moby/moby/issues/3127 for reference

**Pull request merged**